packages/chart/src/components/LineChart/helpers.ts

```diff
@@ -133,75 +172,88 @@ const handleFirstIndex = (data, seriesKey, preliminaryData) => {
   let pairCount = '0'
   const result = {
     data: { '0': [] },
-    style: ''
+    style: '',
+    color: ''
   }
-  let lastAddedIndex = -1 // Tracks the last index added to the result
+  const lastIndexDataItem = data[data.length - 1]
+
+  const dynamicSeriesKey = dynamicCategory ? originalSeriesKey : seriesKey
+  let lastAddedIndex = -1
   preliminaryData?.forEach(pd => {
-    if (pd.type === 'effect') return
-    if (
-      data[data.length - 1][seriesKey] === pd.value &&
-      pd.style &&
-      (!pd.column || pd.column === seriesKey) &&
-      pd.type == 'suppression' &&
-      !pd.hideLineStyle
-    ) {
+    if (pd.type === 'effect') return []
+    if (isSuppressed(pd, lastIndexDataItem)) {
       const lastIndex = data.length - 1
-      const modifiedItem = { ...data[lastIndex], [seriesKey]: 0 }
+      const modifiedItem = { ...data[lastIndex], [dynamicSeriesKey]: 0 }
       result.data[pairCount].push(modifiedItem)
 
       // Find previous calculable item
       let prevIndex = lastIndex - 1
-      while (prevIndex >= 0 && !isCalculable(data[prevIndex][seriesKey])) {
+      while (prevIndex >= 0 && !isCalculable(data[prevIndex][dynamicSeriesKey])) {
         prevIndex--
       }
       if (prevIndex >= 0 && lastAddedIndex !== prevIndex) {
         result.data[pairCount].push(data[prevIndex])
         lastAddedIndex = prevIndex
       }
       result.style = pd.style
+      result.color = colorScale(modifiedItem[dynamicCategory])
     }
   })
 
   return result
 }
 
-function handleMiddleIndices(data, seriesKey, preliminaryData) {
-  // slice data to remove first and last object these no need for handleMiddleIndices
-
+const handleMiddleIndices = ({
+  data,
+  seriesKey,
+  preliminaryData,
+  dynamicCategory,
+  originalSeriesKey,
+  colorScale,
+  isSuppressed
+}) => {
   let result = {
     data: {},
-    style: ''
+    style: '',
+    color: 'red'
   }
   // Variable to count the number of sibling pairs found
   let pairCount = 1
-
+  const dynamicSeriesKey = dynamicCategory ? originalSeriesKey : seriesKey
   // Loop through the data array to find each occurrence of the target value
+
   data.forEach((item, index) => {
     preliminaryData.forEach(pd => {
-      const targetValue = pd.value
-      if (item[seriesKey] === targetValue) {
+      if (isSuppressed(pd, item)) {
         let siblingBefore = null
         let siblingAfter = null
 
         // Find the nearest numeric sibling before the current index
         for (let i = index - 1; i >= 0; i--) {
-          if (isCalculable(data[i][seriesKey])) {
+          if (isCalculable(data[i][dynamicSeriesKey])) {
             siblingBefore = data[i]
             break // Stop searching once a valid sibling is found
           }
         }
 
         // Find the nearest numeric sibling after the current index
         for (let j = index + 1; j < data.length; j++) {
-          if (isCalculable(data[j][seriesKey])) {
+          if (isCalculable(data[j][dynamicSeriesKey])) {
```

Comment:
Throughout the rest of this PR, continue adding checks before accessing values that could potentially be undefined.

Reply:
it is checking here :  const dynamicSeriesKey = dynamicCategory ? originalSeriesKey : seriesKey

Reply:
I added a test to the PR that will show the issue. Can you make sure that gets satisfied?